documentation/source/codes/codes.rst

```diff
@@ -0,0 +1,139 @@
+
+
+    class RunMode(interface.RunMode):
+        RUN = auto()
+
+
+    class Setup(interface.Setup):
+
+        def _run(self):
+            # Write input file
+            pass
+
+
+    class Run(interface.Run):
+
+        def _run(self):
+            self._run_subprocess(self._binary)
+
+
+    class Teardown(interface.Teardown):
+
+        def _run(self):
+            # read from the output file
+            pass
+
+
+    class Solver(interface.FileProgramInterface):
+        _runmode = RunMode
+
+        def __init__(
+            self,
+            params,
+            build_config=None,
+            run_dir: Optional[str] = None,
+        ):
+            super().__init__(
+                "MYPROG",
+                params,
+                build_config.get("mode", "run"),
+                binary=build_config.get("binary", None),
+                run_dir=run_dir,
+                mappings=[],
+                problem_settings=build_config.get("problem_settings", None),
+            )
+
+
+FileProgramInterface
+""""""""""""""""""""
+
+The FileProgramInterface class collects all the tasks together providing a single point to interface between bluemira and the external program.
+A child of FileProgramInterface is the only class that needs to be imported to run a specific solver as seen below.
+
+.. code-block:: python
+
+    import bluemira.codes.mycode as mycode
+
+    params: ParameterFrame
+    build_config: Dict
+
+    solver = mycode.Solver(params, build_config)
+    solver.run()
+
+All mappings for a code are passed to the ParameterFrame from the child class. The RunMode, Setup, Run and Teardown classes are forced to inherit from their respective baseclasses, and a few properties for ease of access are defined. The runmode and the directory in which the code is run are set in the class initialisation.
+
+Variables can be passed through the run method to each task. Using keyword arguments is recommended so that the variables are used correctly.
+
+The only class that technically needs to be defined is RunMode. The rest can be set to None (although they will produce a warning) or just not defined in the inherited class attributes.
+
+RunMode
+"""""""
+
+Each run mode of the code should be defined as a class attribute inherited from this class. The name of the run mode corresponds to the task method that is called when the solver is run. The method that is called is prefixed with '_' and the lower case version of the run mode eg. RunMode.RUN would call '_run' . Tasks do not need to have any run methods. The methods will only be called if they exists.
+
+Tasks
+"""""
+
+The basic task that the three task types inherit from (setup, run, teardown)
+The run subprocess method is defined here as some tasks other than run may want to run an external program. All stdout/err outputs of any external code are captured here so we can control what is outputted to the screen. Stdout is send to the INFO logger and stderr is sent to the ERROR logger.
```

Comment:
`` on the second run, actually makes it easier to read this sentence!

outputted -> output, no?
send -> sent